Rename splitByRanks to agglomerateByRanks and add new option as.list

theotimeee opened the pull request:
This PR renames `splitByRanks` to `agglomerateByRanks` and add a new argument `as.list`.

- If `as.list = TRUE` : The `SimpleList` of agglomerated `SummarizedExperiment` objects is returned by `agglomerateByRanks`
- If `as.list = FALSE` : The `SimpleList` of `SummarizedExperiment` objects is added to `altExps` of the `SummarizedExperiment` passed as a parameter

Commit:
rename splitByRanks to agglomerateByRanks and add option as.list

Commit:
fix example and replace splitByRanks with agglomerateByRanks

Commit:
correct example

Commit:
move documentation

Commit:
up

TuomasBorman commented on R/splitByRanks.R:
```diff
@@ -121,28 +115,48 @@ setGeneric("splitByRanks",
-setMethod("splitByRanks", signature = c(x = "SummarizedExperiment"),
-    function(x, ranks = taxonomyRanks(x), na.rm = TRUE, ...){
+setMethod("agglomerateByRanks", signature = c(x = "SummarizedExperiment"),
+    function(x, ranks = taxonomyRanks(x), na.rm = TRUE, as.list = TRUE, ...){
+        browser()
         args <- .norm_args_for_split_by_ranks(na.rm = na.rm, ...)
-        .split_by_ranks(x, ranks, args)
+        if( !.is_a_bool(as.list) ){
+            stop("'as.list' must be TRUE or FALSE.", call. = FALSE)
+        }
+        if( as.list ){
+            .split_by_ranks(x, ranks, args)
+        }
+        else{
+            altExps(x) <- .split_by_ranks(x, ranks, args)
+            x
+        }
```
SummarizedExpriment does not have altExps slot. You could convert it to TreeSE and give warning.

Also: you are replacing existing altExps with new list (old alternative experiments are lost). Just add these new ones to the list

TuomasBorman commented on R/splitByRanks.R:
```diff
@@ -121,28 +115,48 @@ setGeneric("splitByRanks",
+#' @rdname agglomerate-methods
 #' @export
-setMethod("splitByRanks", signature = c(x = "TreeSummarizedExperiment"),
+setMethod("agglomerateByRanks", signature = c(x = "TreeSummarizedExperiment"),
           function(x, ranks = taxonomyRanks(x), na.rm = TRUE, ...){
               callNextMethod()
           }
```
I think method for TreeSE is not needed because TreeSE is SCE (and that function is called)

TreeSE --> SCE (does have altExp) --> SE

TuomasBorman commented on R/splitByRanks.R:
```diff
@@ -121,28 +115,48 @@ setGeneric("splitByRanks",
-setMethod("splitByRanks", signature = c(x = "SingleCellExperiment"),
-          function(x, ranks = taxonomyRanks(x), na.rm = TRUE, ...){
+setMethod("agglomerateByRanks", signature = c(x = "SingleCellExperiment"),
+          function(x, ranks = taxonomyRanks(x), na.rm = TRUE, as.list = TRUE,
+                    ...){
```
as.list = FALSE is better option (recommended way). In deprecate.R, you can have agglomerateByRanks(as.list = FALSE) hard-coded inside splitByRanks function since the function did not have that parameter

Commit:
convert SE to TSE in agglomerateByRanks SE method

add to AltExps instead of replacing them

Commit:
merge devel

Merge branch 'devel' of https://github.com/microbiome/mia into agglo_all_ranks

# Conflicts:
#	NEWS

Commit:
up

Commit:
deprecate splitByRanks
switch as.list default value at FALSE

Commit:
Merge devel

Commit:
Merge devel

Commit:
merge devel

Merge branch 'devel' of https://github.com/microbiome/mia into agglo_all_ranks

# Conflicts:
#	DESCRIPTION
#	NEWS
#	R/deprecate.R
#	man/deprecate.Rd

Commit:
merge devel

Merge branch 'devel' of https://github.com/microbiome/mia into agglo_all_ranks

# Conflicts:
#	DESCRIPTION
#	NEWS
#	R/agglomerate.R
#	man/agglomerate-methods.Rd

Commit:
merge devel

Merge branch 'devel' of https://github.com/microbiome/mia into agglo_all_ranks

# Conflicts:
#	DESCRIPTION
#	NAMESPACE
#	NEWS
#	R/agglomerate.R
#	R/splitByRanks.R
#	R/splitOn.R
#	man/agglomerate-methods.Rd
#	man/splitByRanks.Rd
#	man/splitOn.Rd
#	tests/testthat/test-5getExperimentCrossAssociation.R

TuomasBorman commented on R/splitByRanks.R:
```diff
@@ -121,37 +116,53 @@ setGeneric("splitByRanks",
-    function(x, ranks = taxonomyRanks(x), na.rm = TRUE, ...){
+setMethod("agglomerateByRanks", signature = c(x = "SummarizedExperiment"),
+    function(x, ranks = taxonomyRanks(x), na.rm = TRUE, as.list = FALSE, ...){
+        browser()
```
browser()

--> Have you run tests? They should notify you about this. (devtools::check() or BiocCheck)

TuomasBorman commented on R/deprecate.R:
```diff
@@ -379,7 +406,7 @@ setMethod("full_join", signature = c(x = "ANY"),
 )
 
-#' @rdname deprecate
+#' @rdname deprecate               
```
Can you revert this line

TuomasBorman commented on R/deprecate.R:
```diff
@@ -362,6 +362,33 @@ loadFromHumann <- function(...) {
     importHUMAnN(...)
 }
 
+#' @rdname deprecate
+#' @export 
+setGeneric("splitByRanks",
+           signature = "x",
+           function(x, ...)
+               standardGeneric("splitByRanks"))
+
+#' @rdname deprecate
+#' @export 
+setMethod("splitByRanks", signature = c(x = "SummarizedExperiment"),
+            function(x,...){
+                .Deprecated(msg = paste0("'splitByRanks' is deprecated. ",
+                                        "Use 'agglomerateByRanks' instead."))
+                agglomerateByRanks(x, as.list = FALSE,...)
+            }
+)
+
+#' @rdname deprecate
+#' @export 
+setMethod("splitByRanks", signature = c(x = "SingleCellExperiment"),
+          function(x,...){
+              .Deprecated(msg = paste0("'splitByRanks' is deprecated. ",
+                                       "Use 'agglomerateByRanks' instead."))
+              agglomerateByRanks(x, as.list = FALSE,...)
+          }
+)
+
```
After considering this, you could leave splitByRanks as an alias.

1. SingleCellExperiment has fnction splitAltExps which is simialr to this
2. We have still unsplitByRanks

--> Add splitByRanks function back, but like it is here: just empty function calling agglomerateByRanks. (Also, only SummarizedExperiment method is needed since SCE is SE)

TuomasBorman commented on R/splitByRanks.R:
```diff
@@ -55,42 +63,29 @@
-#' altExps(GlobalPatterns)
-#' altExp(GlobalPatterns,"Kingdom")
-#' altExp(GlobalPatterns,"Species")
+#' # agglomerateByRanks
+#' # 
+#' x <- agglomerateByRanks(GlobalPatterns)
+#' altExps(x)
+#' altExp(x,"Kingdom")
+#' altExp(x,"Species")
 #'
```
x could be tse

TuomasBorman commented on R/splitByRanks.R:
```diff
@@ -55,42 +63,29 @@
-# splitByRanks
-
-#' @rdname splitByRanks
+#'
```
Add NULL

theotimeee replied:
Here I added the agglomerateByRanks information to agglomerate-methods man page. If I add NULL, I think it does not add all the information written above to the man page and only add agglomerateByRanks methods to Usage part

TuomasBorman replied:
If you add #' @rdname agglomerate-methods to line 83 it should work. If not, then this is ok. This NULL resets roxygen settings --> that way we have full control on where these document chunks are going.

theotimeee replied:
I tried but it did not work, I am not sure but NULL may work as intended only with document chunks using #' @name xyz

TuomasBorman replied:
Might be. Thanks!

TuomasBorman commented on R/splitByRanks.R:
```diff
@@ -121,37 +116,53 @@ setGeneric("splitByRanks",
+        browser()
         args <- .norm_args_for_split_by_ranks(na.rm = na.rm, ...)
-        .split_by_ranks(x, ranks, args)
+        if( !.is_a_bool(as.list) ){
+            stop("'as.list' must be TRUE or FALSE.", call. = FALSE)
+        }
+        if( as.list ){
+            .split_by_ranks(x, ranks, args)
+        }
+        else{
+            altExps(x) <- .split_by_ranks(x, ranks, args)
+            x
+        }
```
Store result to variable and end the function with return(x) --> tidier

TuomasBorman replied:
Also do this:

x <- .split_by_ranks(x, ranks, args)
if( !as.list ){
    # Add to altexps
}
return(x)

TuomasBorman replied:
That is simpler

TuomasBorman replied:
Moreover, you are overwriting existing altExps which is problem. Instead, append the altExps list

TuomasBorman replied:
This is method for SE. It does not have altExps so you have to convert it to TreeSE

theotimeee replied:
> Also do this:
> 
> x <- .split_by_ranks(x, ranks, args) if( !as.list ){ # Add to altexps } return(x)

If I do that, won't the method return a SimpleList no matter the value of `as.list`? The value returned by `.split_by_ranks` internal function is a SimpleList if I am not mistaken. Do we want to return a SimpleList even when the result is added to altExps?

TuomasBorman replied:
Yes, you are correct. How about

```
res <- .split_by_ranks()
# Add to altExp if user has specified to do so
if( !as.list ){
    # Create a general function that adds results to altExps (That could be used in other functions also in the future)
    res <- .add_to_altExps(x, res)
}
return(res)
```

That might be easier to read

TuomasBorman commented on R/splitByRanks.R:
```diff
@@ -121,37 +116,53 @@ setGeneric("splitByRanks",
+                if( !.is_a_bool(as.list) ){
+                    stop("'as.list' must be TRUE or FALSE.", call. = FALSE)
+                }
+                if( as.list ){
+                    .split_by_ranks(x, ranks, args)
+                }
+                else{
+                    x <- as(x, "TreeSummarizedExperiment")
+                    warning("SummarizedExperiment does not have altExps slot. ",
+                            "Therefore, it has been converted to ",
+                            "TreeSummarizedExperiment.")
+                    altExps(x) <- c(altExps(x),.split_by_ranks(x, ranks, args))
```
Instead, can you just call callNextMethod() --> to not copy paste same code

TuomasBorman replied:
SCE already has altExps slot. SE does not have

TuomasBorman commented on NEWS:
```diff
@@ -106,7 +106,8 @@ Changes in version 1.11.x
 + Implemented the setTaxonomyRanks function to specify which ranks are recognized as taxonomy ranks.
 + Rename cluster to addCluster
-+ rename importers loadFromBiom, loadFromQIIME2, readQZA, loadFromMothur, loadFromMetaphlan, loadFromHumann
++ rename importers loadFromBiom, loadFromQIIME2, readQZA, loadFromMothur, 
+loadFromMetaphlan, loadFromHumann
```
Can you revert this line

TuomasBorman commented:
To summarize:

1. Leave splitByRanks as alias. (just empty function calling agglomerateByRanks)
2. In SE method, add TreeSE conversion if the class is SE. (Give warning about the conversion)
3. Add SCE/TreeSE specific functionality to SCE method (TreeSE is SCE). Then call SE method.
4. TreeSE method is just calling SCE method.

Commit:
fix code reviewed

Commit:
up

Commit:
simplify agglomerateByRanks code

TuomasBorman committed:
sync

Sync

# Conflicts:
#	DESCRIPTION
#	NEWS
#	man/agglomerate-methods.Rd

TuomasBorman committed:
sync

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
Merge branch 'devel' into agglo_all_ranks

TuomasBorman committed:
up

TuomasBorman committed:
Merge branch 'devel' into agglo_all_ranks

TuomasBorman committed:
up

TuomasBorman committed:
Merge branch 'agglo_all_ranks' of github.com:microbiome/mia into agglo_all_ranks

TuomasBorman committed:
up

TuomasBorman committed:
up
